### PR TITLE
add clusterwide compaction for dbs

### PR DESCRIPTION
**Pull request by robertkowalski**
COUCHDB-2492

Still missing: per view compaction

### CODE REVIEW

File: src/chttpd_db.erl [hunk 1]
```diff
@@ -225,9 +225,19 @@ maybe_flush_changes_feed(Acc0, Data, Len) ->
...
 
-handle_compact_req(Req, _) ->
-    Msg = <<"Compaction must be triggered on a per-shard basis in CouchDB">>,
-    couch_httpd:send_error(Req, 403, forbidden, Msg).
+handle_compact_req(#httpd{method='POST'}=Req, Db) ->
+    case Req#httpd.path_parts of
+        [_DbName, <<"_compact">>] ->
+            ok = couch_db:check_is_admin(Db),
+            couch_httpd:validate_ctype(Req, "application/json"),
+            _ = couch_httpd:body(Req),
```
Comment (eiri) [inline src/chttpd_db.erl]: If you are not interested in the response, why do you need an assign to pass through? :)

Comment (kxepal) [inline src/chttpd_db.erl]: I believe it's intentional. See https://github.com/apache/couchdb-couch/pull/36 for the case.

File: src/chttpd_db.erl [hunk 2]
```diff
@@ -225,9 +225,19 @@ maybe_flush_changes_feed(Acc0, Data, Len) ->
     },
     {ok, Acc}.
 
-handle_compact_req(Req, _) ->
-    Msg = <<"Compaction must be triggered on a per-shard basis in CouchDB">>,
-    couch_httpd:send_error(Req, 403, forbidden, Msg).
+handle_compact_req(#httpd{method='POST'}=Req, Db) ->
+    case Req#httpd.path_parts of
```
Comment (kxepal) [inline src/chttpd_db.erl]: Why not to match path_parts with the function signature?

**Comment by kxepal** [comment]
Does it works?
```
Compiled src/chttpd_httpd_handlers.erl
/home/kxepal/projects/asf/couchdb/couchdb/src/chttpd/src/chttpd_db.erl:236: syntax error before: 'end'
/home/kxepal/projects/asf/couchdb/couchdb/src/chttpd/src/chttpd_db.erl:17: function handle_compact_req/2 undefined
Compiling /home/kxepal/projects/asf/couchdb/couchdb/src/chttpd/src/chttpd_db.erl failed:
```

File: src/chttpd_db.erl (hunk 1 above)
Comment (eiri) [inline src/chttpd_db.erl]: I understand what call to `body` does, but assign to untagged match all has at most a decorative value.

File: src/chttpd_db.erl (hunk 2 above)
Comment (robertkowalski) [inline src/chttpd_db.erl]: will add another case for view compaction soon

**Comment by robertkowalski** [comment]
> Does it works?

ah good catch, had some leftovers from the changes i already made for view copaction

File: src/chttpd_db.erl [hunk 3]
```diff
@@ -225,9 +225,23 @@ maybe_flush_changes_feed(Acc0, Data, Len) ->
     },
     {ok, Acc}.
 
-handle_compact_req(Req, _) ->
-    Msg = <<"Compaction must be triggered on a per-shard basis in CouchDB">>,
-    couch_httpd:send_error(Req, 403, forbidden, Msg).
+handle_compact_req(#httpd{method='POST'}=Req, Db) ->
+    couch_httpd:validate_ctype(Req, "application/json"),
+    case Req#httpd.path_parts of
```
Comment (rnewson) [inline src/chttpd_db.erl]: this is the first time I've seen path matching within a function, everywhere else we do this at the top level of the function. So, please make this two separate clauses.

Comment (robertkowalski) [inline src/chttpd_db.erl]: interesting, i took the original compact handler at https://github.com/apache/couchdb-couch/blob/aa5d63fce03f6bac18e7c1ffbbaf84f898baf010/src/couch_httpd_db.erl#L161 as an inspiration. will change it

File: src/chttpd_db.erl [hunk 4]
```diff
@@ -225,9 +225,24 @@ maybe_flush_changes_feed(Acc0, Data, Len) ->
     },
     {ok, Acc}.
 
-handle_compact_req(Req, _) ->
-    Msg = <<"Compaction must be triggered on a per-shard basis in CouchDB">>,
-    couch_httpd:send_error(Req, 403, forbidden, Msg).
+handle_compact_req(#httpd{
+        method='POST', path_parts=[_DbName, <<"_compact">>, _DesignName | _]
```
Comment (rnewson) [inline src/chttpd_db.erl]: fix indentation. best to have this on one line even if's long.

**Comment by rnewson** [comment]
+1 after indentation and squash

**Comment by rnewson** [comment]
changing my mind on this to -1, I commented on the jira ticket COUCHDB-2492 with more justification.

**Comment by robertkowalski** [comment]
feel free to revert it

**Comment by robertkowalski** [comment]
its on master already

**Comment by robertkowalski** [comment]
added some comments to the ticket

**Pull request closed without merging**